README.md
# flutter_custom_ui

"Flutter custom ui" is a custom widget that implements a thermometer gauge bar.

## Where to use

It can be used to show a certain mission or achievement rate. 
For example, complete 10 missions of the day gauge 5/10.

- this is sample image.
<img src="assets/sample_image.png" alt="sample image" width="300"/>

## How to use

1. Enter the maximum gauge number and the current gauge number. (necessary)
* Enter the current gauge number not to exceed the maximum gauge number.

2. The current achievement count is displayed on the right side of the gauge bar. (Optional)
   Usage: showCountView: true (defalut: false)

3. For the gauge bar gradation color, input Color values from bottom to top and from left to right.
   (Optional, there is a default value.)
   Usage: gradientStartColor: Colors.blueAccent, gradientEndColor: Colors.yellowAccent,

4. Adjustable gauge bar up and down, left and right length and round size
   (option)

##### Sample code ([Source Code](/lib/sample/sample_code.dart))




CHANGELOG.md
## 1.0.0+1

- flutter custom ui first published
- This version include widget that temperature bar style

test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter_custom_ui/temperature_bar.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    // await tester.pumpWidget(const MyApp());

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


lib/home.dart
import 'dart:math';

import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:flutter_custom_ui/src/temperature_horizontal_bar.dart';
import 'package:flutter_custom_ui/src/temperature_vertical_bar.dart';

class Home extends StatefulWidget {

  const Home({Key? key}) : super(key: key);

  @override
  State<StatefulWidget> createState() => _Home();
}

class _Home extends State<Home> {
  int? _currentIndex;
  int? _currentIndex2;

  @override
  void initState() {

    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: Row(
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            TemperatureVerticalBar(10, _currentIndex ?? 1,
              showCountView: true,
            ),
            TemperatureHorizontalBar(10, _currentIndex2 ?? 1,
              gradientStartColor: Colors.blueAccent,
              gradientEndColor: Colors.yellowAccent,
              showCountView: true,
            )
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: () {
          if (mounted) {
            setState(() {
              _currentIndex = Random().nextInt(11);
              _currentIndex2 = Random().nextInt(11);
            });
          }
        },
        child: const Icon(Icons.play_arrow),
      ),
    );
  }

}

lib/main.dart
import 'dart:math';

import 'package:flutter/material.dart';
import 'package:flutter_custom_ui/sample_code.dart';
import 'package:flutter_custom_ui/src/temperature_horizontal_bar.dart';
import 'package:flutter_custom_ui/src/temperature_vertical_bar.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Custom UI',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Flutter Custom UI'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: const SampleCode(),
    );
  }
}


lib/sample_code.dart
import 'dart:math';

import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:flutter_custom_ui/src/temperature_horizontal_bar.dart';
import 'package:flutter_custom_ui/src/temperature_vertical_bar.dart';

class SampleCode extends StatefulWidget {

  const SampleCode({Key? key}) : super(key: key);

  @override
  State<StatefulWidget> createState() => _SampleCode();
}

class _SampleCode extends State<SampleCode> {
  int? _currentIndex;
  int? _currentIndex2;

  @override
  void initState() {

    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: Row(
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            TemperatureVerticalBar(10, _currentIndex ?? 1,
              showCountView: true,
            ),
            TemperatureHorizontalBar(10, _currentIndex2 ?? 1,
              gradientStartColor: Colors.blueAccent,
              gradientEndColor: Colors.yellowAccent,
              showCountView: true,
            )
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: () {
          if (mounted) {
            setState(() {
              _currentIndex = Random().nextInt(11);
              _currentIndex2 = Random().nextInt(11);
            });
          }
        },
        child: const Icon(Icons.play_arrow),
      ),
    );
  }

}

lib/src/temperature_vertical_bar.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:flutter_custom_ui/src/values/circle_value.dart';
import 'package:flutter_custom_ui/src/values/const_value.dart';

import 'colors/tp_colors.dart';

class TemperatureVerticalBar extends StatefulWidget {
  final int maxIndex;
  final int currentIndex;
  final Color? baseBgColor;
  final Color? gradientBottomColor;
  final Color? gradientTopColor;
  final double? barWidth;
  final double? barHeight;
  final double? circleSize;
  final int? barPointCount;
  final bool showCountView;

  const TemperatureVerticalBar(
    this.maxIndex,
    this.currentIndex,
  {Key? key,
    this.baseBgColor = TPColors.baseBgColor,
    this.gradientBottomColor = TPColors.baseGradientBottomColor,
    this.gradientTopColor = TPColors.baseGradientTopColor,
    this.barWidth,
    this.barHeight,
    this.circleSize,
    this.barPointCount,
    this.showCountView = false,
  }) : super(key: key);

  @override
  State<StatefulWidget> createState() => _TemperatureVerticalBar();
}

class _TemperatureVerticalBar extends State<TemperatureVerticalBar> {
  double? _barWidth;
  double? _barHeight;
  double? _circleSize;

  @override
  void initState() {

    if (mounted) {
      setState(() {
        _barWidth = widget.barWidth ?? ConstValue.baseVerticalWidth;
        _barHeight = widget.barHeight ?? ConstValue.baseVerticalHeight;
        _circleSize = widget.circleSize ?? CircleValue.size;
      });
    }

    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return _buildBody(width: _circleSize!);
  }

  Widget _buildBody({double width = CircleValue.size}) {
    return Wrap(
      children: [
        Row(
          crossAxisAlignment: CrossAxisAlignment.end,
          children: [
            SizedBox(
              width: width * 1.5,
              child: Stack(
                alignment: AlignmentDirectional.bottomCenter,
                children: [
                  Stack(
                    alignment: Alignment.centerLeft,
                    children: [
                      _buildBarBg(width: _barWidth!, height: _barHeight!),
                    ],
                  ),
                  Positioned(
                      bottom: 0,
                      left: 0,
                      right: 0,
                      child: Stack(
                        alignment: AlignmentDirectional.center,
                        children: [
                          _buildCircle(circleSize: _circleSize!),
                          _buildCircle(circleSize: _circleSize! * 0.9, bgColor: widget.gradientBottomColor),
                          _buildCircle(circleSize: _circleSize! / 3, bgColor: widget.baseBgColor)
                        ],
                      )
                  ),
                  Positioned(
                      child: _buildBar(width: _barWidth! ,height: _barHeight! * (widget.currentIndex > widget.maxIndex ? 1 : (widget.currentIndex / widget.maxIndex * 100 / 100)))
                  ),
                  Positioned(
                      left: 0, top: 0, bottom: 0,
                      child: SizedBox(
                        height: (_barHeight ?? ConstValue.baseVerticalHeight) + (_circleSize ?? CircleValue.size),
                        child: Column(
                          mainAxisAlignment: MainAxisAlignment.center,
                          children: _buildBarPointLine(),
                        ),
                      )
                  )
                ],
              ),
            ),
            _buildCount()
          ],
        )
      ],
    );
  }

  _buildCount() {
    return !widget.showCountView ? const SizedBox() : Padding(
        padding: const EdgeInsets.only(bottom: 10,),
        child: Text('${widget.currentIndex}/${widget.maxIndex}'),
    );
  }

  _buildBar({double width = ConstValue.baseVerticalWidth, double height = ConstValue.baseVerticalHeight}) {
    return Column(
      children: [
        Container(
          decoration: BoxDecoration(
            borderRadius: const BorderRadius.only(topLeft: Radius.circular(CircleValue.radius), topRight: Radius.circular(CircleValue.radius)),
            gradient: LinearGradient(colors: [widget.gradientBottomColor!, widget.gradientTopColor!], begin: Alignment.bottomCenter, end: Alignment.topCenter),
          ),
          width: width / 1.8,
          height: height - (height / 2),
        ),
        AnimatedContainer(
          decoration: BoxDecoration(
            color: widget.gradientBottomColor!,
          ),
          margin: EdgeInsets.only(bottom: _circleSize ?? CircleValue.size),
          width: width / 1.8,
          height: height / 2,
          duration: const Duration(seconds: 1),
        ),
      ],
    );
  }

  _buildBarPointLine() {
    List<Widget> list = [];
    int barPointCount = widget.barPointCount ?? ((_barHeight ?? ConstValue.baseVerticalHeight) < 100 ? 4 : 7);

    for (int i=0; i<barPointCount; i++) {
      list.add(Container(
        margin: EdgeInsets.only(top: 3, bottom: 4, left: (_circleSize! - (_barWidth! / 1.8))),
        width: (_barWidth ?? ConstValue.baseVerticalWidth) * 0.25,
        height: 1.5,
        color: widget.baseBgColor!.withOpacity(0.8),
        child: const SizedBox(),
      ));
    }

    return list;
  }

  _buildBarBg({double width = ConstValue.baseVerticalWidth, double height = ConstValue.baseVerticalHeight}) {
    return Column(
      children: [
        Container(
          margin: const EdgeInsets.only(bottom: 10),
          decoration: BoxDecoration(
            borderRadius: const BorderRadius.only(topLeft: Radius.circular(CircleValue.radius), topRight: Radius.circular(CircleValue.radius)),
            border: Border.all(width: 6, color: widget.baseBgColor!)
          ),
          width: width * 0.95,
          height: height + (_circleSize ?? CircleValue.size) - 2,
        ),
      ],
    );
  }

  _buildCircle({double circleSize = CircleValue.size, Color? bgColor}) => Container(
    decoration: BoxDecoration(
      borderRadius: const BorderRadius.all(Radius.circular(CircleValue.radius)),
      color: bgColor ?? widget.baseBgColor,
      border: Border.all(width: bgColor == null ? 5 : 0, color: bgColor ?? widget.baseBgColor!)
    ),
    child: SizedBox(width: circleSize, height: circleSize,),
  );

}

lib/src/temperature_horizontal_bar.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:flutter_custom_ui/src/values/circle_value.dart';
import 'package:flutter_custom_ui/src/values/const_value.dart';

import 'colors/tp_colors.dart';

class TemperatureHorizontalBar extends StatefulWidget {
  final int maxIndex;
  final int currentIndex;
  final Color? baseBgColor;
  final Color? gradientStartColor;
  final Color? gradientEndColor;
  final double? barWidth;
  final double? barHeight;
  final double? circleSize;
  final int? barPointCount;
  final bool showCountView;

  const TemperatureHorizontalBar(
    this.maxIndex,
    this.currentIndex,
      {Key? key,
    this.baseBgColor = TPColors.baseBgColor,
    this.gradientStartColor = TPColors.baseGradientBottomColor,
    this.gradientEndColor = TPColors.baseGradientTopColor,
    this.barWidth,
    this.barHeight,
    this.circleSize,
    this.barPointCount,
    this.showCountView = false,
  }) : super(key: key);

  @override
  State<StatefulWidget> createState() => _TemperatureHorizontalBar();
}

class _TemperatureHorizontalBar extends State<TemperatureHorizontalBar> {
  double? _barWidth;
  double? _barHeight;
  double? _circleSize;

  @override
  void initState() {

    if (mounted) {
      setState(() {
        _barWidth = widget.barWidth ?? ConstValue.baseHorizontalWidth;
        _barHeight = widget.barHeight ?? ConstValue.baseHorizontalHeight;
        _circleSize = widget.circleSize ?? CircleValue.size;
      });
    }

    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return _buildBody(height: _circleSize!);
  }

  Widget _buildBody({double height = CircleValue.size}) {
    return Wrap(
      children: [
        Row(
          crossAxisAlignment: CrossAxisAlignment.end,
          children: [
            SizedBox(
              height: height * 1.5,
              child: Stack(
                alignment: AlignmentDirectional.centerStart,
                children: [
                  Stack(
                    alignment: Alignment.centerLeft,
                    children: [
                      _buildBarBg(width: _barWidth!, height: _barHeight!),
                    ],
                  ),
                  Positioned(
                      bottom: 0,
                      top: 0,
                      left: 0,
                      child: Stack(
                        alignment: AlignmentDirectional.center,
                        children: [
                          _buildCircle(circleSize: _circleSize!),
                          _buildCircle(circleSize: _circleSize! * 0.9, bgColor: widget.gradientStartColor),
                          _buildCircle(circleSize: _circleSize! / 3, bgColor: widget.baseBgColor)
                        ],
                      )
                  ),
                  Positioned(
                      child: _buildBar(height: _barHeight! ,width: _barWidth! * (widget.currentIndex > widget.maxIndex ? 1 : (widget.currentIndex / widget.maxIndex * 100 / 100)))
                  ),
                  Positioned(
                      top: 0, left: 0, right: 0,
                      child: SizedBox(
                        width: (_barWidth ?? ConstValue.baseHorizontalWidth) + (_circleSize ?? CircleValue.size),
                        child: Row(
                          mainAxisAlignment: MainAxisAlignment.center,
                          children: _buildBarPointLine(),
                        ),
                      )
                  )
                ],
              ),
            ),
            _buildCount()
          ],
        )
      ],
    );
  }

  _buildCount() {
    return !widget.showCountView ? const SizedBox() : Padding(
      padding: EdgeInsets.only(bottom: _circleSize! / 2, left: 5),
      child: Text('${widget.currentIndex}/${widget.maxIndex}'),
    );
  }

  _buildBar({double width = ConstValue.baseHorizontalWidth, double height = ConstValue.baseHorizontalHeight}) {
    return Row(
      children: [
        Container(
          decoration: BoxDecoration(
            color: widget.gradientStartColor!,
          ),
          margin: EdgeInsets.only(left: _circleSize ?? CircleValue.size),
          width: width / 2,
          height: height / 1.8,
        ),
        AnimatedContainer(
          decoration: BoxDecoration(
            borderRadius: const BorderRadius.only(bottomRight: Radius.circular(CircleValue.radius), topRight: Radius.circular(CircleValue.radius)),
            gradient: LinearGradient(colors: [widget.gradientStartColor!, widget.gradientEndColor!], begin: Alignment.centerLeft, end: Alignment.centerRight),
          ),
          width: width - (width / 2),
          height: height / 1.8,
          duration: const Duration(seconds: 1),
        ),
      ],
    );
  }

  _buildBarPointLine() {
    List<Widget> list = [];
    int barPointCount = widget.barPointCount ?? ((_barWidth ?? ConstValue.baseHorizontalWidth) < 100 ? 4 : 7);

    for (int i=0; i<barPointCount; i++) {
      list.add(Container(
        margin: EdgeInsets.only(top: (_circleSize! - (_barHeight! / 1.8)), right: 4, left: 5),
        width: 1.5,
        height: (_barHeight ?? ConstValue.baseHorizontalHeight) * 0.25,
        color: widget.baseBgColor!.withOpacity(0.8),
        child: const SizedBox(),
      ));
    }

    return list;
  }

  _buildBarBg({double width = ConstValue.baseHorizontalWidth, double height = (ConstValue.baseHorizontalHeight * 0.95)}) {
    return Row(
      crossAxisAlignment: CrossAxisAlignment.start,
      mainAxisAlignment: MainAxisAlignment.center,
      children: [
        Container(
          margin: const EdgeInsets.only(left: 10),
          decoration: BoxDecoration(
              borderRadius: const BorderRadius.only(bottomRight: Radius.circular(CircleValue.radius), topRight: Radius.circular(CircleValue.radius)),
              border: Border.all(width: 6, color: widget.baseBgColor!)
          ),
          width: width + (_circleSize ?? CircleValue.size) - 5,
          height: height,
        ),
      ],
    );
  }

  _buildCircle({double circleSize = CircleValue.size, Color? bgColor}) => Container(
    decoration: BoxDecoration(
      borderRadius: const BorderRadius.all(Radius.circular(CircleValue.radius)),
      color: bgColor ?? widget.baseBgColor,
      border: Border.all(width: bgColor == null ? 5 : 0, color: bgColor ?? widget.baseBgColor!)
    ),
    child: SizedBox(width: circleSize, height: circleSize,),
  );

}